### PR TITLE
Add customizable main loop iterator proposal code

### PR DESCRIPTION
**Pull request by AndreaCatania**
_PR has the aim to show the code change needed to integrate this proposal: https://github.com/godotengine/godot-proposals/issues/1592_

---

It's possible to customize the [Main::iterator](https://github.com/godotengine/godot/blob/master/main/main.cpp#L2349) from a module by adding some rules on the module `config.py`:

**my_module/config.py**
```python
def can_build(env, platform):
    return True

def configure(env):
    pass

def has_custom_iterator():
    # Optional
    return True

def disable_main_iterator_process():
    # Optional
    return False

def disable_main_iterator_physics():
    # Optional
    return True

def disable_main_iterator_audio():
    # Optional
    return False
```

**has_custom_iterator**
When `has_custom_iterator` returns `True`, the compiler will add the function `custom_iteration` on the `main.cpp` that the module can implement:

**my_module/custom_iterator.cpp**
```cpp
bool Main::custom_iteration(float p_process_delta, float p_physics_delta, MainFrameTime *p_frame_time, float p_time_scale) {

	// This function is called each frame, before godot `process`.

	// Controls whether wants to stop the engine execution.
	return true;
}
```

You can already see it in action here: https://github.com/AndreaCatania/godex/blob/main/main.cpp#L14

**disable_main_iterator_process**, **disable_main_iterator_physics** and **disable_main_iterator_audio**
When these functions, which are optional, returns `True` the specific process within the iterator function is disabled. For example, when `disable_main_iterator_physics` returns `True` the physics is not processed.

This PR, also adds 2 new functions on the `Engine.h` that fully allow to control the engine physics execution from a module.

### CODE REVIEW

File: SConstruct [hunk 1]
```diff
@@ -539,6 +539,9 @@ if selected_platform in platform_list:
                 env.module_icons_paths.append(path + "/" + "icons")
             modules_enabled[name] = path
 
+            if getattr(config, "has_custom_iterator", False) and config.has_custom_iterator():
+                env.Append(CPPDEFINES=["CUSTOM_ITERATOR"])
```
Comment (Xrayez) [inline SConstruct:543]: Some other module may already define this. Not sure if crucial but shouldn't hurt anything.
```suggestion
                env.AppendUnique(CPPDEFINES=["CUSTOM_ITERATOR"])
```

**Comment by Xrayez** [comment]
The above suggestion may signify that this should be a SCons option, like `scons use_custom_iterator=yes`. But AFAIK, custom defines can be already added using existing options:

https://github.com/godotengine/godot/blob/97f116d36bdc8a335f26eebb70a7480d96536ecb/SConstruct#L168-L175

File: SConstruct (hunk 1 above)
Comment (AndreaCatania) [reply SConstruct:543]: I can add it, yes.

**Comment by AndreaCatania** [comment]
Done in this way you can define modules that automatically override the `Main::iterator` and other modules that give to the user the possibility to chose it. For example:
- You can read the passed arguments into the `config.py` and so make the function `has_custom_iterator` return `False`. Allowing the user to choose when to use the custom iterator or not.
- In the same way it's possible to just return `True` when the module must always override the `Main::iterator`.

So in this way you have more possibilities, making it more versatile.

Also, I don't expect that you will ever have 2 modules into the same project that want to override the iterator. You may have it, but I'm quite sure that you will activate one or the other, at compile time.

**Comment by AndreaCatania** [comment]
Updated this PR with the new iterator customization approach.

**Comment by AndreaCatania** [comment]
I've updated this PR, to improve the naming, and I wrote about my experience using it, in these six month in Godex, here: https://github.com/godotengine/godot-proposals/issues/1592#issuecomment-817161648. I'll take care to document this feature.

**Comment by ValorZard** [comment]
Has there been any movement on this issue? Seems like the solution has been found here, but maybe this PR has just been lost in the hubbub of 3.3 and 4.0 update

**Comment by AndreaCatania** [comment]
@ValorZard Reduz wants to review this after the new GDNative mechanism is fully implemented, so no movements yet but we are still working on this.

**Comment by neikeq** [comment]
I think it should be possible to change this at runtime, so maybe with the introduction of the new GDNative mechanism this can be reconsidered.

**Pull request closed without merging**